File: src/BOJ/Tree/PP_11725.java
package BOJ.Tree;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class PP_11725 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        ArrayList<ArrayList<Integer>> list = new ArrayList<>();
        int[] answer = new int[n + 1];
        for (int i = 0; i <= n + 1; i++) {
            list.add(new ArrayList<>());
        }

        for (int i = 1; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            list.get(a).add(b);
            list.get(b).add(a);
        }

        bfs(1, list, answer, n);

        for (int i = 2; i < answer.length; i++) {
            System.out.println(answer[i]);
        }
    }

    private static void bfs(int start, ArrayList<ArrayList<Integer>> list, int[] answer, int n) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        answer[start] = 1;

        while (!queue.isEmpty()) {
            int parent = queue.poll();

            for (int item : list.get(parent)) {
                if (answer[item] == 0) {
                    answer[item] = parent;
                    queue.offer(item);
                }
            }
        }
    }
}

File: src/BOJ/Tree/PP_11725.kt
package BOJ.Tree

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import kotlin.math.pow

val answer = ArrayList<MutableList<String>>()

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val k = br.readLine().toInt()
    val list = br.readLine().split(" ")

    for (i in 0 until k) {
        answer.add(mutableListOf())
    }

    findNum(0, list.size - 1, 0, k, list)

    for (i in 0 until k) {
        for (j in 0 until answer[i].size)
            bw.write(answer[i][j] + " ")
        bw.write("\n")
    }

    bw.flush()
    bw.close()
}

fun findNum(start: Int, end: Int, floor: Int, k: Int, list: List<String>) {
    if (floor == k) {
        return
    }

    val mid = (start + end) / 2
    answer[floor].add(list[mid])

    findNum(start, mid - 1, floor + 1, k, list)
    findNum(mid + 1, end, floor + 1, k, list)
}

